CSharp/Library/Microsoft.Bot.Builder/Dialogs/PromptDialog.cs

```diff
@@ -577,39 +635,59 @@ public PromptChoice(IEnumerable<T> options, string prompt, string retry, int att
 
+            /// <summary>   Constructor for a prompt choice dialog. </summary>
+            /// <param name="choices">Dictionary of the options to choose from with the synonyms.</param>
+            /// <param name="prompt">   The prompt. </param>
+            /// <param name="retry">    What to display on retry. </param>
+            /// <param name="attempts"> Maximum number of attempts. </param>
+            /// <param name="promptStyle"> Style of the prompt <see cref="PromptStyle" /> </param>
+            /// <param name="descriptions">Descriptions to show for each option.</param>
+            public PromptChoice(IDictionary<T, IEnumerable<T>> choices, string prompt, string retry, int attempts, PromptStyle promptStyle = PromptStyle.Auto, IEnumerable<string> descriptions = null)
+                : this(new PromptOptions<T>(prompt, retry, options: choices.Keys.ToList(), attempts: attempts, promptStyler: new PromptStyler(promptStyle), descriptions: descriptions?.ToList()), choices)
+            {
+            }
+            
             /// <summary>
             /// Constructs a choice dialog.
             /// </summary>
             /// <param name="promptOptions"> The prompt options</param>
-            public PromptChoice(PromptOptions<T> promptOptions)
+            /// <param name="choices">Dictionary of the options to choose from with the synonyms.</param>
+            public PromptChoice(PromptOptions<T> promptOptions, IDictionary<T, IEnumerable<T>> choices = null)
                 : base(promptOptions)
             {
                 SetField.CheckNull(nameof(promptOptions.Options), promptOptions.Options);
-            }
-
-            public virtual Tuple<bool, int> ScoreMatch(T option, string input)
-            {
-                var trimmed = input.Trim();
-                var text = option.ToString();
-                bool occurs = text.IndexOf(trimmed, StringComparison.CurrentCultureIgnoreCase) >= 0;
-                bool equals = text == trimmed;
-                return occurs
-                    ? Tuple.Create(equals, trimmed.Length)
-                    : null;
+                if (choices == null)
+                {
+                    this.choices = new Dictionary<T, IEnumerable<T>>();
+                    promptOptions.Options.ToList().ForEach(x => this.choices.Add(x, null));
+                }
+                else
+                {
+                    this.choices = choices;
+                }
             }
 
             protected override bool TryParse(IMessageActivity message, out T result)
             {
                 if (!string.IsNullOrWhiteSpace(message.Text))
                 {
-                    var scores = from option in this.promptOptions.Options
-                                 let score = ScoreMatch(option, message.Text)
-                                 select new { score, option };
+                    var entityMatches = PromptRecognizers.RecognizeChoices(message.Text, choices);
```

Comment:
The promptConfirm does not use the resources directly. The synonyms must be generated from the `patterns` property